text/204-golang-bindings.md

```diff
@@ -443,34 +453,180 @@ wrappers, e.g.
+            Cluster:                args.Cluster,
+            DesiredCount:           args.DesiredCount,
+            ServiceName:            args.ServiceName,
+            MaxHealthyPercent:      args.MaxHealthyPercent,
+            MinHealthyPercent:      args.MinHealthyPercent,
+            HealthCheckGracePeriod: args.HealthCheckGracePeriod,
+            CloudMapOptions:        args.CloudMapOptions,
+            PropagateTags:          args.PropagateTags,
+            EnableECSManagedTags:   args.EnableECSManagedTags,
+            DeploymentController:   args.DeploymentController,
+        }),
+        launchType: args.LaunchType,
+    }
+}
+```
+
+The advantage of this approach is that adding a new optional property to a super struct no longer results in a breaking change (the field will be
+absent from the subtype until it is re-generated against the new parent), as the embedding technique guarantees methods are promoted from the
+supertype. It also guarantees immutability of built instances, as the actual implementation is not exported.
+
+The inconvenient is that this adds relatively heavy boilerplate around use of structs, where the name of the struct is repeated twice in sequence.
+This could lead to some user confusion (at least until the suer becomes familiar with this idiom). A typical call would look like so:
+
+```go
+// Pretending one can instantiate the BaseService construct directly
+NewBaseService(scope, "ID", NewBaseServiceProps(BaseServiceProps{
+    // Note: optional fields omitted (as a user could decide to do)
+    Cluster: cluster,
+}))
+```
+
 ### Notes/Concerns
 
 * Like the AWS Go SDK, we can use pointers to primitive types to simulate that fields are optional (i.e. allow null values rather than default "empty"
   values for each type, which are not nullable). However, this might result in a less than ideal developer experience. Another option is to use a
-wrapper type for optional values.
-* Generated Go interfaces corresponding to a datatype interface would need a suffix, e.g. Iface, in order to disambiguate it from the struct. This is
-  a bit verbose, and it may be worth considering switching the naming (i.e. adding a suffix to the struct instead) if the interface name is what will
-primarily be used by the customer.
+  wrapper type for optional values.
+* Generated Go interfaces corresponding to a datatype interface would need a name addendum, e.g. I prefix of Iface suffix, in order to disambiguate
+  it from the struct name. This can be confusing (I prefix makes structs and behavioral interfaces similarly named on APIs) or a bit verbose (Iface
+  suffix), and it may be worth considering switching the naming (i.e. adding a suffix to the struct instead), as the interface name is what will
+  primarily be used by the customer.
+* As the `@jsii/kernel` may decide to pass a data type by-reference (and not by-value), either a concrete struct needs to be implemented that does
```

Comment:
what's the use case for the kernel to pass data by-ref, isn't the whole point of data structs that they are always passed by value?

Reply:
The by-ref behavior is actually expected/used in Java & C# in the current state of things, IIRC to work around serialisation issues that otherwise arised. The issue down the line is that changing this now would likely be risky, and likely a large-is effort. It'll likely be much easier to deal with this quirk, at least in the short term.